lib/modules/manager/bazel-module/context.ts

```diff
@@ -0,0 +1,150 @@
+import {
+  ArrayFragment,
+  AttributeFragment,
+  BooleanFragment,
+  ChildFragments,
+  Fragment,
+  RecordFragment,
+  StringFragment,
+  ValueFragment,
+  asFragment,
+  safeAsValue,
+} from './fragments';
+import { Stack } from './stack';
+
+// Represents the fields that the context must have.
+export interface CtxCompatible {
+  results: ValueFragment[];
+  stack: Stack<Fragment>;
+}
+
+export class Ctx implements CtxCompatible {
+  results: ValueFragment[];
+  stack: Stack<Fragment>;
+
+  constructor(results: ValueFragment[] = [], stack = Stack.create<Fragment>()) {
+    this.results = results;
+    this.stack = stack;
+  }
+
+  // This exists because the good-enough-parser gives a cloned instance of our
+  // Ctx. It is missing the Ctx prototype. This function adds the proper prototype
+  // to the context and the items referenced by the context.
+  static as(obj: CtxCompatible): Ctx {
+    if (obj instanceof Ctx) {
+      return obj;
+    }
+    const results = obj.results.map((frag) => RecordFragment.as(frag));
+    const stackItems = obj.stack.map((frag) => asFragment(frag));
+    return new Ctx(results, Stack.create(...stackItems));
+  }
```

Comment:
this is expensive to do in every parsing step because of the recreation of a lot of objects.

@zharinov Do you have an idea how to better handle the parsing without require to clone and recreate so many objects?

Reply:
Context object is meant to be just a plain data structure, i.e. arrays and fields with primitive types, everything you can serialize as JSON, and everything you can describe with `Record<T>` and `T[]`. I think, encapsulating it to the own class is the unnecessary abstraction here, though with this claim I'm stepping on the territory of professional preferences and beliefs.

Reply:
Fixed at `1.1.23`

Reply:
@zharinov Do you have a PR to update `renovate` to `1.1.22`? If not, I can put one up.

Reply:
I will remove Ctx.as once 1.1.23 for `good-enough-parser` is available in `renovate`.

Reply:
@zharinov Do we have an ETA for the upgrade to 1.1.23?

Reply:
@viceice Could you please check the box in the Dashboard?

Reply:
Done